enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+result of the migration, result in system rolling back to previous deployment.
+
+Decision flow describing whether to block or attempt data migration can be summarized as:
+- If version metadata is missing, assume 4.13
+- Refuse to migrate if
+  - Version of data is present on list of prohibited migrations
+  - Version skew between data on disk and binary is bigger than X.Y+1
+  - Y version of binary is older than Y version of data
+- Skip migration if X.Y are the same
+- Attempt data migration otherwise
+
+#### Staging new ostree commits on top of unhealthy systems
+
+Automated handling of MicroShift data in case of unhealthy system is
+complicated due to ambiguity regarding admin's intention and each option has
+different trade offs without clear winner.
+
+For that reason, when needing to stage a new commit on top of unhealthy system
+one of two actions must be performed:
+- system must be brought to healthy state, or
+- MicroShift data should be deleted completely resulting in clean start which
+  is the same as first run.
+
+Only exception is when rollback commit doesn't feature MicroShift.
+This special case handles scenarios when device is preinstalled with system
+without MicroShift, then commit with MicroShift runs, but is unhealthy so it
+rolls back to factory system, later another commit with MicroShift is staged
+and runs, but it should not be held back by stale data.
+
+### Open Questions [optional]
+
+### Workflows in detail
+
+#### Decision tree
+
+###### Backup management
+
+- "Data" refers to MicroShift's data
+- Terms such as _previous boot_, _previous boot's commit_, and
+  _previous commit with MicroShift_ are related to MicroShift's "history of commits"
+  (therefore only updated when commit runs MicroShift).
+- _Current boot's commit_ is what currently runs on the device
+- _Rollback_ refers to ostree's rollback, i.e. older commit that will be booted
+  if newly staged commit is unhealthy.
+- Not covered "else" conditions will result in error and not starting the cluster.
+
+---
+
+1. If data does not exist
+   - MicroShift is running for the first time on the device.
+   - There's nothing to backup or migrate, skip to running cluster.
+
+1. If data exists but is missing version metadata:
+   Assume it's 4.13, back up the data and update "history of commits" (use `4.13` as ID),
+   and proceed to [data migration](#data-migration-1).
+
+1. If previous boot was healthy:
+   **backup data for previous boot**, check if next point is applicable, proceed to [data migration](#data-migration-1).
+
+1. (Regardless of previous boot health)<a name="current-commit-different-restore"></a>
+   If current boot's commit is different from previous boot's commit and backup exist for current commit:
+   **restore data for current commit, skip data migration, [start cluster](#starting-the-cluster)**.
+   > Existence of backup for "current boot's commit" means commit already ran on the device and was healthy.
```

Comment:
Actually it's both. Both after staging and rolling back (manually on-demand or automatically via greenboot), _previous boot's commit_ will differ from _current boot's commit_.
Basically it's "we booted into different commit that we have backup for? restore"

Reply:
Also take a look at line 466:
```
1. If backup for _current boot's commit_ exists: **restore**
```

Effectively flow looks like: if previous boot was healthy - backup, then consider whether to restore or do something else.
BTW. if we reboot to the same commit, it looks (on paper) like we'd backup and immediately restore what we just backed up [0]. While it's "technically the truth" I would expect the implementation to skip the restore.

[0] Hmm, that made me think that someone could think restore would remove the backup, but it's not the case (at least for me)

Reply:
Since we backup/restore/migrate on each boot (rather than on new commits), do you foresee scenario when "rebooted from unhealthy system" should result in keeping the data as is?
FYI Current decision tree very strongly leans on assumption that upgrades happen only from healthy system. So if there's a decision path not covered in this version of the document (compared to previous very verbose and explicit), implementation should "refuse to proceed and cause a roll back"

Reply:
The original plan did call for us to always backup if the previous boot had been healthy, so we always have the latest data in case we need to roll back. In the first few discussions, we were only going to keep 1 total backup. Now we're keeping 1 per commit. I'm not sure we need more copies than that. Do you see a benefit of doing that that I might be missing?